smart_card_connector_app/src/mock-chrome-api.js

```diff
@@ -0,0 +1,205 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * @fileoverview This file contains a mock of chrome.smartCardProviderPrivate
+ * API.
+ */
+
+goog.provide('GoogleSmartCard.ConnectorApp.MockChromeApi');
+
+goog.require('goog.Disposable');
+goog.require('goog.asserts');
+goog.require('goog.testing');
+goog.require('goog.testing.MockControl');
+goog.require('goog.testing.PropertyReplacer');
+goog.require('goog.testing.StrictMock');
+goog.require('goog.testing.mockmatchers');
+
+goog.setTestOnly();
+
+goog.scope(function() {
+
+/**
+ * Contains the names of all events exposed by chrome.smartCardProviderPrivate
+ * API.
+ */
+const EXTENSION_API_EVENTS = [
+  'onEstablishContextRequested',
+  'onReleaseContextRequested',
+  'onListReadersRequested',
+  'onGetStatusChangeRequested',
+  'onCancelRequested',
+  'onConnectRequested',
+  'onDisconnectRequested',
+  'onTransmitRequested',
+  'onControlRequested',
+  'onGetAttribRequested',
+  'onSetAttribRequested',
+  'onStatusRequested',
+  'onBeginTransactionRequested',
+  'onEndTransactionRequested',
+];
+
+/**
+ * Contains the names of all functions exposed by
+ * chrome.smartCardProviderPrivate API.
+ */
+const EXTENSION_FUNCTIONS = [
+  'reportEstablishContextResult',
+  'reportReleaseContextResult',
+  'reportListReadersResult',
+  'reportGetStatusChangeResult',
+  'reportPlainResult',
+  'reportConnectResult',
+  'reportDataResult',
+  'reportStatusResult',
+];
+
+/**
+ * Contains the ResultCode enum values exposed by
+ * chrome.smartCardProviderPrivate API.
+ */
+const RESULT_CODES = [
+  'SUCCESS',
+  'REMOVED_CARD',
+  'RESET_CARD',
+  'UNPOWERED_CARD',
+  'UNRESPONSIVE_CARD',
+  'UNSUPPORTED_CARD',
+  'READER_UNAVAILABLE',
+  'SHARING_VIOLATION',
+  'NOT_TRANSACTED',
+  'NO_SMARTCARD',
+  'PROTO_MISMATCH',
+  'SYSTEM_CANCELLED',
+  'NOT_READY',
+  'CANCELLED',
+  'INSUFFICIENT_BUFFER',
+  'INVALID_HANDLE',
+  'INVALID_PARAMETER',
+  'INVALID_VALUE',
+  'NO_MEMORY',
+  'TIMEOUT',
+  'UNKNOWN_READER',
+  'UNSUPPORTED_FEATURE',
+  'NO_READERS_AVAILABLE',
+  'SERVICE_STOPPED',
+  'NO_SERVICE',
+  'COMM_ERROR',
+  'INTERNAL_ERROR',
+  'UNKNOWN_ERROR',
+  'SERVER_TOO_BUSY',
+  'UNEXPECTED',
+  'SHUTDOWN',
+  'UNKNOWN'
+];
+
+/**
+ * Sets up mocks for
+ * chrome.smartCardProviderPrivate.*.addListener()/removeListener(),
+ * report*Result functions. Checks that all events have a subscriber and that
+ * this subscriber is removed properly. Checks that all the expectations are
+ * satisfied on disposal.
+ */
+GoogleSmartCard.ConnectorApp.MockChromeApi = class extends goog.Disposable {
+  /**
+   * @param {!goog.testing.PropertyReplacer} propertyReplacer
+   */
+  constructor(propertyReplacer) {
+    super();
+    /** @type {Object.<string, Function>} */
+    this.listeners_ = {};
+    this.listenersMockControl_ = new goog.testing.MockControl();
```

Comment:
Usually there's one `MockControl` for the whole test. It makes sure there's one central place that verifies the mock expectations and nothing is forgotten.

So can we keep creating `mockControl` in the test file and pass it here as a parameter?

Reply:
done